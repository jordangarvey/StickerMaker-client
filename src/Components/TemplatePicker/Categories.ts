/**
 * Enum defining the categories that a sticker can be edited by
 */
enum Categories {
	Colour = "colour",
	Shape = "shape",
	Text = "text"
}

export default Categories;
