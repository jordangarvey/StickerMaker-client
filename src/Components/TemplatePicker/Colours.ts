/**
 * Enum defining the colours that a sticker can be
 */
enum Colours {
	Black = "#000000",
	Grey1 = "#333333"
}

export default Colours;
