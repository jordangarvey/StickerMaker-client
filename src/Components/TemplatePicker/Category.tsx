import React, { FC } from "react";
import { StyleSheet, Text, TouchableOpacity } from "react-native";

import Categories from "./Categories";

import * as C from "../../Global/Colours";


/**
 * Interface for the Category props
 */
interface ICategoryProps {
	/** The currently selected category this option belongs to e.g. shapes, text */
	category: Categories;
	/** On press callback to be fired whenever the option button is pressed */
	onPress(category: Categories): void;
	/** Optional boolean for whether or not this option is currently selected */
	selected?: boolean;
}

/**
 * Component to render a Category button
 */
const Category: FC<ICategoryProps> = (props) => {
	return (
		<TouchableOpacity
			style={[
				styles.category,
				props.selected && { backgroundColor: C.primary }
			]}
			onPress={props.onPress.bind(null, props.category)}
		>
			<Text style={props.selected && { color: C.secondary }}>{props.category.charAt(0).toUpperCase() + props.category.slice(1)}</Text>
		</TouchableOpacity>
	);
};

/**
 * Styles for the Category button
 */
const styles = StyleSheet.create({
	category: {
		alignItems: "flex-end",
		backgroundColor: C.secondary,
		borderRadius: 6,
		height: 80,
		flexDirection: "row",
		justifyContent: "center",
		marginHorizontal: 20,
		padding: 5,
		width: 60
	}
});

export default Category;
