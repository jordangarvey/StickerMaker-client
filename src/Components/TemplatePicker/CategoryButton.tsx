import React, { FC } from "react";
import { StyleSheet, Text, TouchableOpacity } from "react-native";

import Categories from "./Categories";
import Shapes from "./Shapes";

import * as C from "../../Global/Colours";


/**
 * Interface for the Category Button props
 */
interface ICategoryButtonProps {
	/** The currently selected category this option belongs to e.g. shapes, text */
	category: Categories;
	/** The option that this button should display e.g. circle, square */
	option: Shapes;
	/** On press callback to be fired whenever the option button is pressed */
	onPress(option: Shapes): void;

	/** Optional boolean for whether or not this option is currently selected */
	selected?: boolean;
}

/**
 * Component to render a Category Button button
 */
const CategoryButton: FC<ICategoryButtonProps> = (props) => {
	return (
		<TouchableOpacity
			onPress={props.onPress.bind(null, props.option)}
			style={[
				styles.CategoryButton,
				props.selected && { backgroundColor: C.primary }
			]}
		>
			<Text style={props.selected && { color: C.textPrimary }}>
				{ props.option.charAt(0).toUpperCase() + props.option.slice(1) }
			</Text>
		</TouchableOpacity>
	);
};

/**
 * Styles for the Category Button button
 */
const styles = StyleSheet.create({
	CategoryButton: {
		alignItems: "flex-end",
		backgroundColor: C.secondary,
		borderRadius: 6,
		height: 80,
		flexDirection: "row",
		justifyContent: "center",
		marginHorizontal: 10,
		padding: 5,
		width: 60
	}
});

export default CategoryButton;
