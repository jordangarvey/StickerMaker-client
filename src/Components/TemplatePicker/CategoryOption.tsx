import React, { FC } from "react";
import { StyleSheet, Text, View } from "react-native";

import Categories from "./Categories";
import Shapes from "./Shapes";

import * as C from "../../Global/Colours";


/**
 * Interface for the Category Option props
 */
interface ICategoryOptionProps {
	/** The currently selected category this option belongs to e.g. shapes, text */
	category: Categories;
	/** The option that this button should display e.g. circle, square */
	option: Shapes;
	/** On press callback to be fired whenever the option button is pressed */
	onPress(option: Shapes): void;
	/** Optional boolean for whether or not this option is currently selected */
	selected?: boolean;
}

/**
 * Component to render a Category Option button
 */
const CategoryOption: FC<ICategoryOptionProps> = (props) => {
	return (
		<View
			onPress={props.onPress.bind(null, props.option)}
			style={[
				styles.categoryOption,
				props.selected && { backgroundColor: C.primary, color: C.textPrimary }
			]}
		>
			<Text>{props.option.charAt(0).toUpperCase() + props.option.slice(1)}</Text>
		</View>
	);
};

/**
 * Styles for the Category Option button
 */
const styles = StyleSheet.create({
	categoryOption: {
		alignItems: "flex-end",
		backgroundColor: C.secondary,
		borderRadius: 6,
		height: 80,
		flexDirection: "row",
		justifyContent: "center",
		marginHorizontal: 10,
		padding: 5,
		width: 60
	}
});

export default CategoryOption;
