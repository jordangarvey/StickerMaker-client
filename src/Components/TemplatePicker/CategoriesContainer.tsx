import React, { FC } from "react";
import { ScrollView, StyleSheet, View } from "react-native";

import Categories from "./Categories";

import Category from "./Category";


/**
 * Interface for the Categories Container props
 */
interface ICategoriesContainerProps {
	/** Callback to be fired whenever the option button is pressed */
	onPress(category: Categories): void;

	/** The currently selected category */
	currentCategory?: Categories;
}

const CategoriesContainer: FC<ICategoriesContainerProps> = (props) => (
	<View style={styles.categoriesContainer}>
		<ScrollView horizontal={true}>
			{
				Object.keys(Categories).map((category: any) => (
					<Category
						key={Categories[category]}
						category={Categories[category] as Categories}
						onPress={props.onPress}
						selected={Categories[category] === props.currentCategory}
					/>
				))
			}
		</ScrollView>
	</View>
);

/**
 * Styles for the Categories Container
 */
const styles = StyleSheet.create({
	categoriesContainer: {
		marginBottom: 20,
		width: "100%"
	}
});

export default CategoriesContainer;
