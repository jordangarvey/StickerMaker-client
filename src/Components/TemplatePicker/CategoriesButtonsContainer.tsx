import React, { FC } from "react";
import { ScrollView, StyleSheet, View } from "react-native";

import categories from "../../Data/categories";

import CategoryButton from "./CategoryButton";


const CategoriesButtonsContainer: FC = () => (
	<View style={styles.categoriesContainer}>
		<ScrollView contentContainerStyle={styles.categoriesScroller} horizontal={true}>
			{
				Object.keys(categories).map(category => (
					<CategoryButton
						key={category}
						category={category}
					/>
				))
			}
		</ScrollView>
	</View>
);

/**
 * Styles for the Categories Buttons Container
 */
const styles = StyleSheet.create({
	categoriesContainer: {
		width: "100%"
	},
	categoriesScroller: {
		flexDirection: "row",
		flexWrap: "wrap",
		justifyContent: "space-evenly",
	}
});

export default CategoriesButtonsContainer;
