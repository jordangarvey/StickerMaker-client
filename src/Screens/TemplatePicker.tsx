import React, { Component } from "react";
import { NativeModules, StyleSheet, View } from "react-native";

import ViewShot from "react-native-view-shot";

import Categories from "../Components/TemplatePicker/Categories";
import Colours from "../Components/TemplatePicker/Colours";
import Shapes from "../Components/TemplatePicker/Shapes";

import ISelectedOptions from "../Components/TemplatePicker/ISelectedOptions";

import TemplateItem from "../Components/TemplatePicker/TemplateItem";
import CategoriesContainer from "../Components/TemplatePicker/CategoriesContainer";
import OptionsContainer from "../Components/TemplatePicker/OptionsContainer";

import * as C from "../Global/Colours";


/**
 * Interface for the Template Picker screen state
 */
interface ITemplatePickerState {
	/** The currently selected options */
	currentOptions: ISelectedOptions;

	/** The currently selected category, if any */
	currentCategory?: Categories;
	/** Optional error that’s thrown from the component */
	error?: Error;
}

/**
 * Component to render the Template Picker screen
 */
class TemplatePicker extends Component<any, ITemplatePickerState> {
	public constructor(props: any) {
		super(props);

		// Intitialise the state
		this.state = {
			currentOptions: {
				colour: Colours.Black,
				shape: Shapes.Rect,
				text: ""
			}
		};

		// Bind this component’s methods
		this.capture = this.capture.bind(this);
		this.onCategoryButtonPress = this.onCategoryButtonPress.bind(this);
		this.onTextValueChange = this.onTextValueChange.bind(this);
		this.onOptionButtonPress = this.onOptionButtonPress.bind(this);
	}

	private async capture() {
		let addBase64Image: string;
		try {
			base64Image = await this.refs.templateItem.capture();
		} catch(error) {
			console.error(error);
			this.setState({ error });
			return;
		}

		NativeModules.BetterClipboard.addBase64Image(addBase64Image);
	}
	
	/**
	 * Method to be fired whenever a category button is pressed
	 */
	private onCategoryButtonPress(category: Categories) {
		const currentCategory = this.state.currentCategory !== category ? category : undefined;
		this.setState({ currentCategory });
	}

	/**
	 * Method to be fired whenever an option button is pressed
	 */
	private onOptionButtonPress(category: Categories, option: Colours | Shapes) {
		const currentOptions = this.state.currentOptions;
		switch(category) {
			case Categories.Colour:
				currentOptions.colour = option as Colours;
				break;
			case Categories.Shape:
				currentOptions.shape = option as Shapes;
				break;
		}

		this.setState({ currentOptions });
	}

	/**
	 * Method to be fired whenever the text value is changed
	 */
	private onTextValueChange(text: string) {
		const currentOptions = this.state.currentOptions;
		currentOptions.text = text;

		this.setState({ currentOptions });
	}

	public render() {
		return (
			<View style={styles.fullWidthBackground}>
				<View style={styles.templateItemContainer}>
					<ViewShot ref="templateItem" options={{ result: "base64" }}>
						<TemplateItem
							colour={this.state.currentOptions.colour}
							shape={this.state.currentOptions.shape}
							text={this.state.currentOptions.text}
						/>
					</ViewShot>
				</View>

				<View style={styles.bottomContainer}>
					<CategoriesContainer
						currentCategory={this.state.currentCategory}
						onPress={this.onCategoryButtonPress}
					/>
					<OptionsContainer
						currentCategory={this.state.currentCategory}
						currentOptions={this.state.currentOptions}
						onCapture={this.capture}
						onChange={this.onTextValueChange}
						onPress={this.onOptionButtonPress}
					/>
				</View>
			</View>
		);
	}
}

const styles = StyleSheet.create({
	bottomContainer: {
		bottom: 0,
		flexDirection: "column",
		justifyContent: "space-between",
		position: "absolute",
		width: "100%"
	},
	fullWidthBackground: {
		...StyleSheet.absoluteFillObject,
		backgroundColor: C.background
	},
	templateItemContainer: {
		alignSelf: "center",
		top: 200
	}
});

export default TemplatePicker;
