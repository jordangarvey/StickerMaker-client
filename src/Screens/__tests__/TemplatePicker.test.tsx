import React from "react";

import renderer from "react-test-renderer";

import TemplatePicker from "../TemplatePicker";


describe("The Template Picker componentâ€™s snapshot", () => {
	it("renders as expected", () => {
		const component = renderer.create(
			<TemplatePicker/>
		).toJSON();


		expect(component).toMatchSnapshot();
	});
});
