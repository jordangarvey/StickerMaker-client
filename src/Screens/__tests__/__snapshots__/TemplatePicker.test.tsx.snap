// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Template Picker componentâ€™s snapshot renders as expected 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5F5F5",
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "top": 200,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#000000",
          },
          Object {
            "alignItems": "center",
            "borderRadius": 6,
            "justifyContent": "center",
          },
          Object {
            "height": 100,
            "width": 200,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "white",
          }
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "bottom": 0,
        "flexDirection": "column",
        "justifyContent": "space-between",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 20,
          "width": "100%",
        }
      }
    >
      <RCTScrollView
        horizontal={true}
      >
        <View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "flex-end",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 6,
                "flexDirection": "row",
                "height": 80,
                "justifyContent": "center",
                "marginHorizontal": 20,
                "opacity": 1,
                "padding": 5,
                "width": 60,
              }
            }
          >
            <Text
              style={false}
            >
              Colour
            </Text>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "flex-end",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 6,
                "flexDirection": "row",
                "height": 80,
                "justifyContent": "center",
                "marginHorizontal": 20,
                "opacity": 1,
                "padding": 5,
                "width": 60,
              }
            }
          >
            <Text
              style={false}
            >
              Shape
            </Text>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "flex-end",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 6,
                "flexDirection": "row",
                "height": 80,
                "justifyContent": "center",
                "marginHorizontal": 20,
                "opacity": 1,
                "padding": 5,
                "width": 60,
              }
            }
          >
            <Text
              style={false}
            >
              Text
            </Text>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#000000",
          "borderWidth": 0.5,
          "height": 300,
          "width": "100%",
        }
      }
    >
      <RCTScrollView
        horizontal={true}
        style={
          Object {
            "height": 200,
            "width": "100%",
          }
        }
      >
        <View />
      </RCTScrollView>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#18A0FB",
            "borderBottomLeftRadius": 30,
            "borderBottomRightRadius": 30,
            "borderTopLeftRadius": 6,
            "borderTopRightRadius": 6,
            "flexDirection": "row",
            "height": 70,
            "justifyContent": "center",
            "marginHorizontal": 20,
            "opacity": 1,
            "padding": 5,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Export to camera roll
        </Text>
      </View>
    </View>
  </View>
</View>
`;
